core/app/models/spree/order_shipping.rb

```diff
@@ -70,7 +70,10 @@ def ship(inventory_units:, stock_location:, address:, shipping_method:,
       # TODO: make OrderShipping#ship_shipment call Shipment#ship! rather than
       # having Shipment#ship! call OrderShipping#ship_shipment. We only really
       # need this `update_columns` for the specs, until we make that change.
-      shipment.update_columns(state: 'shipped', shipped_at: Time.current)
+      attributes = { state: "shipped", shipped_at: Time.current }
+      shipment.update_columns(attributes)
+      # Make sure the in-memory shipments of the order have the correct state for OrderUpdate#determine_shipment_state
+      @order.shipments.detect { |s| shipment.id == s.id }&.assign_attributes(attributes)
```

Comment:
A few lines above we use `update!`, so I don't see why we need to go with `update_columns` here. I understand that we want the changes from these inventory unit shipments to be propagated to the order's shipments, but I think we can accomplish that much easier by doing

```
shipment.update!(state: :shipped, shipped_at: Time.current)
@order.shipments.reset
```

The advantage of that approach would be: definitely no changed shipments in the order updater, and more understandability of this code for people looking at it :) What do you think?

Reply:
We cannot be certain that there are no changes on the in-memory shipments we gather from the inventory units (that potentially could be from other orders).

If @solidusio/core-team agrees that we take that risk I would love to got with this approach instead.

Reply:
We should trust the test suite. Otherwise, there's no point in it :man_shrugging: Do you think we need to add a more specific test?

Reply:
I made a test locally with the following code (improvable for sure):

```ruby
if shipment.inventory_units.reload.all? { |iu| iu.shipped? || iu.canceled? }
  shipment.update!(state: 'shipped', shipped_at: Time.current, tracking: tracking_number)
else
  shipment.update!(tracking: tracking_number)
end
```

and made some tests with `binding.irb` right after this execution. Apparently, at this point, the value of the shipment associated to the `@order` (taken with`@order.shipments.detect { |s| shipment.id == s.id }`) is already updated with the value just set with the `update!` call. No reload needed and, if I got it correctly, without the reload we won't have any risk to clear existing changes on the in-memory shipment. isn't it?

Reply:
Tried that.

Unfortunately the `order.shipment_state` is still `"partial"`, unless I also call

```rb
@order.shipments.reload
```

right before `@order.recalculate`

I pushed an update. Tell me what you think @kennyadsl @mamhoff @jarednorman @waiting-for-dev